### PR TITLE
Handle packages.config restore scenario where multiple projects with the same name exist in the same solution

### LINKED ISSUE #13465: Visual Studio and PMC restore/update fails when multiple packages.config projects in the solution share the same name (An item with the same key has already been added)
https://github.com/NuGet/Home/issues/13456

```code
PM> Update-Package -reinstall
Update-Package : An item with the same key has already been added.
At line:1 char:1
+ Update-Package -reinstall
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Update-Package], ArgumentException
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.UpdatePackageCommand
 
PM> $error[0].Exception.StackTrace
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at NuGet.PackageManagement.PackageRestoreManager.<GetRestoreAuditProperties>d__31.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.PackageRestoreManager.<RestoreMissingPackagesAsync>d__35.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.PowerShellCmdlets.NuGetPowerShellBaseCommand.<CheckMissingPackagesAsync>d__64.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.PowerShellCmdlets.PackageActionBaseCommand.<<Preprocess>b__39_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at NuGet.PackageManagement.PowerShellCmdlets.PackageActionBaseCommand.Preprocess()
   at NuGet.PackageManagement.PowerShellCmdlets.UpdatePackageCommand.Preprocess()
   at NuGet.PackageManagement.PowerShellCmdlets.UpdatePackageCommand.ProcessRecordCore()
   at NuGet.PackageManagement.PowerShellCmdlets.NuGetPowerShellBaseCommand.ProcessRecord()
```

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/13465

Regression? Last working version:

## Description

See comment in the code for details. 
This is a selective limitation being introduced here as the risk of the right fix is too high.
https://github.com/NuGet/Home/issues/13466 covers the long term fix. 

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled - https://github.com/NuGet/Home/issues/13467
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by nkolev92** [comment]
Working on getting a test for this: 

VS disallows loading projects with the same name, so I'm surprised this is even a thing.

File: test/TestUtilities/Test.Utility/PackageManagement/TestSolutionManager.cs [hunk 1]
```diff
@@ -79,24 +79,27 @@ public TestSolutionManager(string solutionDirectory)
             SolutionDirectory = solutionDirectory;
         }
 
-        public MSBuildNuGetProject AddNewMSBuildProject(string projectName = null, NuGetFramework projectTargetFramework = null, string packagesConfigName = null)
+        public MSBuildNuGetProject AddNewMSBuildProject(string projectName = null, NuGetFramework projectTargetFramework = null, string projectPath = null, bool validateExistingProject = true)
         {
-            var existingProject = Task.Run(async () => await GetNuGetProjectAsync(projectName));
-            existingProject.Wait();
-            if (existingProject.IsCompleted && existingProject.Result != null)
+            if (validateExistingProject)
```
Comment (nkolev92) [inline test/TestUtilities/Test.Utility/PackageManagement/TestSolutionManager.cs:84]: Currently the validation just checks the project name.
In practice, projects are stored via 4 different keys, include project id, and project paths. 

The primary key however is full name, which is a project path. 

I think these APIs need refactored, but I don't want to get to that now given that we want to service these changes:

https://github.com/NuGet/Client.Engineering/issues/2855

**Review by zivkan — APPROVED** [review]

**Comment by sanderdelfos** [comment]
> Working on getting a test for this:
> 
> VS disallows loading projects with the same name, so I'm surprised this is even a thing.

@nkolev92 
 Fyi, I experienced this problem in a VS solution that contained two projects called 'Services'. One was a C++ project (.vcxproj), the other a C# (.csproj). VS does allow this, never had any problems with it. 
For now, we solved the nuget restore by renaming one of the two.

**Comment by nkolev92** [comment]
> > Working on getting a test for this:
> > VS disallows loading projects with the same name, so I'm surprised this is even a thing.
> 
> @nkolev92 Fyi, I experienced this problem in a VS solution that contained two projects called 'Services'. One was a C++ project (.vcxproj), the other a C# (.csproj). VS does allow this, never had any problems with it. For now, we solved the nuget restore by renaming one of the two.

Thanks for responding @sanderdelfos. 

Someone pointed out in one of the linked issues that they need to in different "solution folders".

**Pull request merged**